base/applications/rapps/gui.cpp

```diff
@@ -257,6 +257,154 @@ class CAppRichEdit:
+        {
+            ::MoveWindow(RichEdit->m_hWnd, 0, 0, GET_X_LPARAM(lParam), GET_Y_LPARAM(lParam), TRUE);
+            break;
+        }
+        case WM_COMMAND:
+        {
+
+            OnCommand(wParam, lParam);
+            break;
+        }
+        case WM_NOTIFY:
+        {
+            NMHDR* NotifyHeader = (NMHDR*)lParam;
+            if (NotifyHeader->hwndFrom == RichEdit->m_hWnd)
+            {
+                switch (NotifyHeader->code)
+                {
+                case EN_LINK:
+                    OnLink((ENLINK*)lParam);
+                    break;
+                }
+            }
+            break;
+        }
+        }
+
+        return FALSE;
+    }
+
+public:
+
+    CAppRichEdit * RichEdit;
+
+    static ATL::CWndClassInfo& GetWndClassInfo()
+    {
+        DWORD csStyle = CS_VREDRAW | CS_HREDRAW;
+        static ATL::CWndClassInfo wc =
+        {
+            {
+                sizeof(WNDCLASSEX),
+                csStyle,
+                StartWindowProc,
+                0,
+                0,
+                NULL,
+                NULL,
+                NULL,
+                (HBRUSH)(COLOR_BTNFACE + 1),
+                NULL,
+                L"RAppsAppInfo",
+                NULL
+            },
+            NULL, NULL, IDC_ARROW, TRUE, 0, _T("")
+        };
+        return wc;
+    }
+
+    HWND Create(HWND hwndParent)
+    {
+        RECT r = { 0,0,0,0 };
+
+        return CWindowImpl::Create(hwndParent, r, L"", WS_CHILD | WS_VISIBLE);
+    }
+
+    BOOL ShowAvailableAppInfo(CAvailableApplicationInfo* Info)
+    {
+        return RichEdit->ShowAvailableAppInfo(Info);
+    }
+
+    BOOL ShowInstalledAppInfo(PINSTALLED_INFO Info)
+    {
+        return RichEdit->ShowInstalledAppInfo(Info);
+    }
+
+    VOID SetWelcomeText()
+    {
+        RichEdit->SetWelcomeText();
+    }
+
+    VOID OnLink(ENLINK* Link)
+    {
+        switch (Link->msg)
+        {
+        case WM_LBUTTONUP:
+        case WM_RBUTTONUP:
+        {
+            if (pLink) HeapFree(GetProcessHeap(), 0, pLink);
+
+            pLink = (LPWSTR)HeapAlloc(GetProcessHeap(), 0,
+                (max(Link->chrg.cpMin, Link->chrg.cpMax) -
+                    min(Link->chrg.cpMin, Link->chrg.cpMax) + 1) * sizeof(WCHAR));
+            if (!pLink)
+            {
+                /* TODO: Error message */
+                return;
+            }
+
+            RichEdit->SendMessageW(EM_SETSEL, Link->chrg.cpMin, Link->chrg.cpMax);
+            RichEdit->SendMessageW(EM_GETSELTEXT, 0, (LPARAM)pLink);
+
+            ShowPopupMenuEx(m_hWnd, m_hWnd, IDR_LINKMENU, -1);
+        }
+        break;
+        }
+    }
+
+    VOID OnCommand(WPARAM wParam, LPARAM lParam)
+    {
+        WORD wCommand = LOWORD(wParam);
+
+        switch (wCommand)
+        {
+        case ID_OPEN_LINK:
+            
+            ShellExecuteW(m_hWnd, L"open", pLink, NULL, NULL, SW_SHOWNOACTIVATE);
+            HeapFree(GetProcessHeap(), 0, pLink);
+			pLink = NULL;
```

Comment:
You seem to have introduced tabs here, instead of spaces

Reply:
ah yes. my bad habit. corrected.

Reply:
You should consider this: https://stackoverflow.com/questions/30140595/show-whitespace-characters-in-visual-studio-code 😄

Reply:
> You should consider this: https://stackoverflow.com/questions/30140595/show-whitespace-characters-in-visual-studio-code 😄

got it. actually I'm using Visual Studio and I've just opened the option to insert spaces instead of tabs. hope I will no longer make this mistake.